### PR TITLE
This will allow the user to filter users by certain criteria

### PR DESCRIPTION
**Pull request by james00100**
## PR Title

Issue Number(s): #105

What does this PR change and why?

Allows for users to filter for other users based on criteria for better matching

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [X] Code follows design and style guidelines
- [X] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

- Database change/migration to run
- Environment config change
- Breaking API change

### Related PRs

- #number_of_pr

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

File: web/db/actions/User.ts [hunk 1]
```diff
@@ -59,6 +73,50 @@ async function updateUserByUid(
...
+  size?: Size[],
+  preferredBreeds?: Breed[],
+  gender?: Gender[],
+  age?: Age[],
+  temperament?: Temperament[],
+  dogsNotGoodWith?: GoodWith[],
+  medical?: Medical[],
+  behavioral?: Behavioral[],
+  houseTrained?: Trained,
+  spayNeuterStatus?: Trained,
```
Comment (katsukixyz) [inline web/db/actions/User.ts]: Use `Status`

File: web/db/actions/User.ts [hunk 2]
```diff
@@ -59,6 +73,50 @@ async function updateUserByUid(
...
+  session?: ClientSession
+): Promise<IUser[] | null> {
```
Comment (katsukixyz) [inline web/db/actions/User.ts]: Take in an object with these keys instead. Session should be a separate parameter, however

File: web/db/actions/User.ts [hunk 3]
```diff
@@ -59,11 +72,58 @@ async function updateUserByUid(
...
+  if (searchParams.size) filter.size = { $all: searchParams.size };
+  if (searchParams.preferredBreeds) filter.preferredBreeds = { $all: searchParams.preferredBreeds };
+  if (searchParams.gender) filter.gender = { $all: searchParams.gender };
+  if (searchParams.age) filter.age = { $all: searchParams.age };
+  if (searchParams.temperament) filter.temperament = { $all: searchParams.temperament };
+  if (searchParams.dogsNotGoodWith) filter.dogsNotGoodWith = { $all: searchParams.dogsNotGoodWith };
+  if (searchParams.medical) filter.medical = { $all: searchParams.medical };
+  if (searchParams.behavioral) filter.behavioral = { $all: searchParams.behavioral };
+  if (searchParams.houseTrained) filter.houseTrained = searchParams.houseTrained;
+  if (searchParams.spayNeuterStatus) filter.spayNeuterStatus = searchParams.spayNeuterStatus;
```
Comment (katsukixyz) [inline web/db/actions/User.ts]: Make a function that takes in `searchParams` and outputs the filter query. In this function, make a map for how each field should end up in the filter query:
```js
{
  role: searchParams.role,
  type: { $all: searchParams.preferredBreeds }, 
  ...
}
```
In the function, iterate over all the fields in searchParams (using a reduce function), adding each value of the map to the filter query object. Call this helper function from `searchUsers` to get the filter query you use.

File: web/server/routers/user.ts [hunk 4]
```diff
@@ -161,4 +162,35 @@ export const userRouter = router({
...
+            spayNeuterStatus: z.nativeEnum(Status).optional(),
+          }),
+        })
+      )
+      .query(async ({ input }) => {
+        try {
+          const { searchParams } = input;
+          const res = await searchUsers(searchParams);
+          return { data: res };
+        } catch (e) {
```
Comment (katsukixyz) [inline web/server/routers/user.ts:189]: Add conditional to check whether `e` is an instance of `TRPCError` - if it is, throw e, else you can throw the internal server error

File: web/server/routers/user.ts [hunk 5]
```diff
@@ -161,4 +162,35 @@ export const userRouter = router({
...
+            size: z.array(z.nativeEnum(Size)).optional(),
+            preferredBreeds: z.array(z.nativeEnum(Breed)).optional(),
+            gender: z.array(z.nativeEnum(Gender)).optional(),
+            age: z.array(z.nativeEnum(Age)).optional(),
+            temperament: z.array(z.nativeEnum(Temperament)).optional(),
+            dogsNotGoodWith: z.array(z.nativeEnum(GoodWith)).optional(),
+            medical: z.array(z.nativeEnum(Medical)).optional(),
+            behavioral: z.array(z.nativeEnum(Behavioral)).optional(),
+            houseTrained: z.nativeEnum(Status).optional(),
+            spayNeuterStatus: z.nativeEnum(Status).optional(),
```
Comment (katsukixyz) [inline web/server/routers/user.ts]: Use `z.object().partial()` instead

**Review by katsukixyz — COMMENTED** [review]
Great work! A few minor comments, but looking great!!

File: web/server/routers/user.ts (hunk 5 above)
Comment (katsukixyz) [reply web/server/routers/user.ts]: Bumping this

**Review by katsukixyz — COMMENTED** [review]
Looks good - just bumped a comment form an earlier review. I'll approve as soon as you're done with that change!

**Review by katsukixyz — APPROVED** [review]
Awesome work! Feel free to merge it in.
